### ISSUE TITLE
GameLift: Unable To Search Sessions via Scripting

### ISSUE THREAD

**Issue by AMZN-Gene** [o3de/o3de#2615]
While there are script nodes to search for gamelift sessions, there's no way to provide search request input

**To Reproduce**
1. Open script canvas
2. Add a "SearchSession" node
3. Notice there's no way to provide a SearchSessionsRequest input

**Expected behavior**
Users should be able to create SearchSessionsRequest variables in order to provide input to the SearchSessions node.
Instead, users are only provided with AWSGameLiftSearchSessionRequest variables which are incompatible
![image](https://user-images.githubusercontent.com/32776221/127533285-908f6638-196a-4766-a768-fc6577d481c2.png)

**Comment by onecent1101** [comment]
This is a feature not supported by ScriptCanvas.
If you check the source code of SearchSessions event, the input is reference of SearchSessionRequest. AWSGameLiftSearchSessionsRequest is a derived class of SearchSessionRequest.
The workaround is to use slot connection NOT drop on slot.
![slotconnection](https://user-images.githubusercontent.com/5900509/127936316-894e1e39-07f3-4b08-a41e-09aef506de1f.PNG)

**Comment by AMZN-Gene** [comment]
The same applied to JoinSession. In my test case, I already know which session I wanted to join, therefore I wanted to create a new ScriptCanvas AWSGameLiftJoinSessionRequest.  But AWSGameLiftJoinSessionRequest isn't accepted as input into JoinSession node.
I recommend providing a JoinSession node which accepts AWSGameLiftJoinSessionRequest _or_ allow users to create the base class  JoinSessionRequest.  In my case for testing, I ended up creating a new behavior method called JoinSessionViaAWSGameLiftJoinSessionRequest().  This way I was able to pass in my hardcoded test  AWSGameLiftJoinSessionRequest

**Comment by onecent1101** [comment]
As scriptcanvas doesn't support drop derived type on slot (connecting to slot works fine), assign to scripting team for further investigation.

**Comment by lmbr-pip** [comment]
@greerdv - Can you provide an update why this is not a sig/content scripting issue? Unclear why this was assigned back to sig-network without comment.

**Comment by lmbr-pip** [comment]
Seems like we need input from scripting, to cover:
* If this is known issue (sounds like it is), if so can we get a link to GHI to track or documentation?
* Or if not known issue, can you confirm this is an issue to as it appears that a derived class is not working as expected.

Reassigning back to sig-content for triage as issue appears to be at scripting level and nothing to do with sig-network aside from it was found on sig-network scripting components.

**Comment by AMZN-Gene** [comment]
If scripting doesn't plan to allow passing in derived type, then the networking/aws team could explicitly adds the new type.

**Comment by onecent1101** [comment]
@AMZN-Gene see my screenshot, it should be clear that you can make connection as expected. scripting supports derived type, the issue is scripting team has inconsistent type check on slot connection and drop on slot as reference

**Comment by monroegm** [comment]
@lsemp3d Can you help with the prioritization of this one?

**Comment by lsemp3d** [comment]
This is a request for the AWS scripting api, @lmbr-pip any thoughts?

**Comment by lsemp3d** [comment]
Moved it to networking and set it to needs-triage for visibility, can you schedule some time to discuss this issue? it appears there is a question on how to develop the API to handle a specific use-case, it may need a hands-on session to form a recommendation.

**Comment by lsemp3d** [comment]
ok, I did some local testing and I found that adding the following code in `Gems\ScriptCanvas\Code\Include\ScriptCanvas\Core\Slot.cpp`:

```
AZ::SerializeContext* serializeContext = nullptr;
        AZ::ComponentApplicationBus::BroadcastResult(serializeContext, &AZ::ComponentApplicationRequests::GetSerializeContext);
        AZ_Assert(serializeContext, "Serialization context not available");

        bool supported = false;
        auto baseClassVisitorFn = [thisType, &supported](const AZ::SerializeContext::ClassData* reflectedBase, const AZ::TypeId& /*rttiBase*/)
        {
            if (!reflectedBase)
            {
                return false; // stop iterating
            }

            supported = (reflectedBase->m_typeId == thisType.GetAZType());
            if (supported)
            {
                return false; // we have a base, stop iterating
            }

            return true; // keep iterating
        };

        AZ::EntityUtils::EnumerateBaseRecursive(serializeContext, baseClassVisitorFn, dataType.GetAZType());

        if (supported)
        {
            return AZ::Success();
        }
```

just before the last line `return AZ::Failure(AZStd::string::format("%s is not a type match for %s", ScriptCanvas::Data::GetName(GetDataType()).c_str(), ScriptCanvas::Data::GetName(dataType).c_str()));`

Will allow the type to be set on the node. However, I'm not setup to test a GameLift session locally, so I don't know whether any issues will popup at runtime.

**Comment by onecent1101** [comment]
![derivedtype](https://user-images.githubusercontent.com/5900509/144335043-1994aa1d-5dd6-4852-9412-bf31c43a7022.PNG)

@lsemp3d Tested with the fix, it can work now, and get expected log after running game.
```
(Script) - fleetid: fleet1 sessionname: testfromscriptcanvas maxplayer: 2.0000 aliasid:
(AWSGameLiftCreateSessionActivity) - Requesting CreateGameSession against Amazon GameLift service ...
(AWSGameLiftCreateSessionActivity) - Built CreateGameSessionRequest with CreatorId=, Name=testfromscriptcanvas, IdempotencyToken=, GameProperties=[], AliasId=, FleetId=fleet1 and MaximumPlayerSessionCount=2
(AWSGameLiftCreateSessionActivity) - CreateGameSession request against Amazon GameLift service is complete
(Script) - arn:aws:gamelift:local::gamesession/fleet1/gsess-ef608cc1-8420-4fc8-a680-209b80d2ff8e
(MultiplayerSample) - CreateSessionAsync complete with result: arn:aws:gamelift:local::gamesession/fleet1/gsess-ef608cc1-8420-4fc8-a680-209b80d2ff8e
```

But there is a minor issue: with the fix, the slot won't be auto converted to reference when you try to drop variable on it. You have to manually convert slot to reference, then you can drop variable.

**Comment by lsemp3d** [comment]
Excellent, are you able to submit a PR with that fix and open an issue for the problem with references?

**Comment by onecent1101** [comment]
Yes, I can submit a PR for fixing it. But I have a question, if you check code above your recommended solution:
```
if (thisType.IS_A(dataType))
{
    return AZ::Success();
}
```

And seems like IS_A is the function should do the type check which includes derived type check as well. (at least based on this function comment)
Plus there are two type match function:
```
AZ::Outcome<void, AZStd::string> Slot::IsTypeMatchFor(const Slot& otherSlot) const
AZ::Outcome<void, AZStd::string> Slot::IsTypeMatchFor(const ScriptCanvas::Data::Type& dataType) const
```
I guess the first one is used for slot connection (as it is working for derived type), and it uses the same IS_A check, why does it work there? should we put that lambda function of derived type check against serialize context inside IS_A?

**Comment by lsemp3d** [comment]
IS_A is only checking for an exact match and not doing a polymorphic check using RTTI

**Comment by onecent1101** [comment]
But function says it supports derived type check
https://github.com/o3de/o3de/blob/development/Gems/ScriptCanvas/Code/Include/ScriptCanvas/Data/Data.h#L130-L131

And why slot connection works for derived type with current implementation?

**Comment by lmbr-pip** [comment]
SIG triage moved to close issue as no longer relevant as we should approach this setup in a different way.

Will ensure update to GameLift functionality support session searching.

**Comment by AMZN-Gene** [comment]
Reopening... customer has the same issue.
